stdlib.sh

```diff
@@ -986,6 +986,38 @@ layout_pipenv() {
   export VIRTUAL_ENV
 }
 
+# Usage: layout poetry
+#
+# Similar to layout_python, but uses Poetry [1] to build a virtualenv from the
+# pyproject.toml [2] located in the same directory.
+#
+# Poetry by default will create a virtual environment under
+# {cache-dir}/virtualenvs where cache-dir equates to $XDG_CACHE_HOME/pypoetry
+# or use the {project-dir}/.venv directory when one is available.
+#
+# [1]: https://python-poetry.org/
+# [2]: https://peps.python.org/pep-0518/
+#
+layout_poetry() {
+    PYPROJECT_TOML="${PYPROJECT_TOML:-pyproject.toml}"
+    if [[ ! -f "$PYPROJECT_TOML" ]]; then
+        log_status "No pyproject.toml found. Executing \`poetry init\` to create a \`$PYPROJECT_TOML\` first."
+        poetry init
+    fi
+
+    VIRTUAL_ENV=$(poetry env info --path 2>/dev/null ; true)
```

Comment:
> `poetry` CLI because it's so slow

Indeed, I even suggested in the [wiki](https://github.com/direnv/direnv/wiki/Python#poetry) to use the `--no-interaction` option with the `init` command (i.e. `poetry --no-interaction init`) to get a basic `pyproject.toml` rather than have the user answer the prompts to create it and risk `direnv` complain the process is taking too long for its liking.

I've not found `poetry install` to be slow but I really just have it install the active python version to the virtual environment and add my projects dependencies or dev dependencies later with `poetry add`.
 
> It might be possible to adapt it for the false case as well, but I haven't invested any time towards that.

Which false case do you mean here. The case where the `virtualenvs.in-project` setting is `false` but the `.venv` exists or where the setting is `false` and the virtual environment exists in `{cache-dir}/virtualenvs`. Would the former case cause issues for poetry in knowing where to install packages? The latter case may be difficult as `poetry` names the virtual environment based on what the user has named their project directory.

Reply:
> In my own custom version of `layout_poetry()` I've added a short-circuiting branch here to avoid invoking the `poetry` CLI because it's so slow.

This is a great idea. We also use .venv in projects, so this has fixed my issue, as I was just digging into why this step was slow and thought it was the poetry env step. I think it's a great tip, so I added it to the wiki as a note.

Reply:
Maybe include it in the `layout_poetry` function in the wiki directly? I can update the PR in the same way.

Reply:
OK, PR updated to reflect this check.

Reply:
I'm not sure I understand this check for `.venv`.  This is very idiosyncratic.  Some people call it `venv`; others call it `ENV`.  See https://github.com/github/gitignore/blob/4488915eec0b3a45b5c63ead28f286819c0917de/Python.gitignore#L123-L127

Also, direnv's stdlib makes no mention of what seems like an arbitrary choice of `.venv`. Here's what's used instead: https://github.com/direnv/direnv/blob/7608dae27e0ca9a88785195afc9a2e0281eb2dc9/stdlib.sh#L827

It seems this would be mixing different layout models, no?

Reply:
In fact, it might make sense to support a custom `VIRTUAL_ENV` that's already defined, as in:
https://github.com/direnv/direnv/blob/7608dae27e0ca9a88785195afc9a2e0281eb2dc9/stdlib.sh#L822-L826

Reply:
> I'm not sure I understand this check for `.venv`. This is very idiosyncratic. Some people call it `venv`; others call it `ENV`.

The check is on `.venv` because that is what `poetry` will look for when the virtual environment is in-project. See [here](https://python-poetry.org/docs/configuration/#virtualenvsin-project).

> It seems this would be mixing different layout models, no?

Not really. One can either allow `direnv` or `poetry` to manage the virtual environments. There will be no conflict but it is probably best practice to be consistent in your choice to avoid confusion.